### PR TITLE
Vite migration

### PR DESCRIPTION
**Pull request by flomillot**
Replace nwb by vite.js. Few changes were necessary.

To make the demo works : 
- change extension file from .js into .jsx when returning react elements
- fix few imports
- add an index.html
- use svgr plugin for SVG and update imports
- use a plugin to fix react-virtualized bad import and update the version to works with the plugin

To make the tests works : 
- add a jest config
- use babel presets

To make the library works when used with CRA : 
- use a plugin to externalize the deps

Additional notes :
- add TypeScript support
- we don't add `type: module` in the package.json and force the output file to take the .js extension to avoid imports compatibility issues when importing this library with other projects using webpack, ie : 

   - accept not fully specified path : https://github.com/webpack/webpack/issues/11467 (mostly for MUI imports)

   - use of Babel interpretation for default exports : https://github.com/webpack/webpack/issues/17288 https://esbuild.github.io/content-types/#default-interop (issue with default import of react-request-fullscreen)

### CODE REVIEW

**Review by jonenst — APPROVED** [review]

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]

**Comment by Tristan-WorkGH** [comment]
Also from <https://vitejs.dev/guide/features#typescript>, it seem that vite won't do a check with `tsc` like CRA.  
Would need to include typescript checks in eslint or babel-eslint pipeline, vite plugin don't seem good.

File: vite.config.mts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        eslint(),
+        dts(),
+        svgr({ include: '**/*.svg' }), // default is { include: "**/*.svg?react" }
+        reactVirtualized(),
+        externalizeDeps({
+            include: [/^react-is(?:\/.*)?$/, /^@mui\/styled-engine(?:\/.*)?$/],
+        }),
+    ],
+
+    build: {
```
Comment (Tristan-WorkGH) [inline vite.config.mts:33]: Have found how to keep file files compiled separately in bundle:
```suggestion
    build: {
        rollupOptions: {
            output: {
                preserveModules: true,
                entryFileNames: '[name].js',
            },
        },
```

Result: 
image

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]

File: vite.config.mts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2024, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import react from '@vitejs/plugin-react';
+import type { PluginOption } from 'vite';
+import { defineConfig } from 'vite';
```
[the remark this answers is not in the document]
Comment (flomillot) [reply vite.config.mts:10]: Done

File: package.json [hunk 3]
```diff
@@ -6,22 +6,17 @@
...
-        "prepublishOnly": "npm run build",
-        "start": "npm install;nwb serve-react-demo",
-        "test": "nwb test-react",
-        "test:coverage": "nwb test-react --coverage",
-        "test:watch": "nwb test-react --server"
+        "watch": "vite watch",
+        "build": "npm install; vite build; npm prune --production",
+        "test": "jest",
+        "start": "vite demo/ --config vite.config.mts",
+        "lint": "eslint . --ext js,jsx,ts,tsx"
```
[the remark this answers is not in the document]
Comment (flomillot) [reply package.json:19]: Added prepublish, eslint in an other PR

File: vite.config.mts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        svgr({ include: '**/*.svg' }), // default is { include: "**/*.svg?react" }
+        reactVirtualized(),
+        externalizeDeps({
+            include: [/^react-is(?:\/.*)?$/, /^@mui\/styled-engine(?:\/.*)?$/],
+        }),
+    ],
+
+    build: {
+        lib: {
+            entry: resolve(__dirname, 'src/index.js'),
```
[the remark this answers is not in the document]
Comment (flomillot) [reply vite.config.mts:35]: Done

File: vite.config.mts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import react from '@vitejs/plugin-react';
+import type { PluginOption } from 'vite';
+import { defineConfig } from 'vite';
+import * as path from 'path';
+import { resolve } from 'path';
```
[the remark this answers is not in the document]
Comment (flomillot) [reply vite.config.mts:12]: Done

File: vite.config.mts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import eslint from 'vite-plugin-eslint';
+import dts from 'vite-plugin-dts';
+import * as fs from 'fs/promises';
+import * as url from 'url';
```
[the remark this answers is not in the document]
Comment (flomillot) [reply vite.config.mts:16]: Done

File: vite.config.mts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            },
+            formats: ['es'],
+        },
+        minify: false, // easier to debug on the apps using this lib
+    },
+});
+
+// Workaround for react-virtualized with vite
+// https://github.com/bvaughn/react-virtualized/issues/1632#issuecomment-1483966063
+function reactVirtualized(): PluginOption {
```
[the remark this answers is not in the document]
Comment (Tristan-WorkGH) [reply vite.config.mts:50]: not mandatory

File: vite.config.mts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import * as url from 'url';
+import { createRequire } from 'node:module';
+import svgr from 'vite-plugin-svgr';
+import { externalizeDeps } from 'vite-plugin-externalize-deps';
+
+export default defineConfig({
+    plugins: [
+        react(),
+        eslint(),
+        dts(),
```
[the remark this answers is not in the document]
Comment (flomillot) [reply vite.config.mts:25]: Will do an other PR and remove dts plugin

File: vite.config.mts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    build: {
+        lib: {
+            entry: resolve(__dirname, 'src/index.js'),
+            name: 'Commons ui',
+            fileName: () => {
+                // from https://github.com/vitejs/vite/discussions/1736#discussioncomment-4997467
+                // in this way, we can have a .js output file without having to add 'type: module' in the package.json which introduce breaking changes
+                return `commons-ui.js`;
+            },
+            formats: ['es'],
```
[the remark this answers is not in the document]
Comment (flomillot) [reply vite.config.mts:42]: Nope

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    "strict": true,
+    "skipLibCheck": true,
+    "esModuleInterop": true,
+    "forceConsistentCasingInFileNames": true,
+    "allowSyntheticDefaultImports": true,
+    "noFallthroughCasesInSwitch": true,
+    "moduleResolution": "node"
+  },
+  "include": ["src"]
+}
```
[the remark this answers is not in the document]
Comment (flomillot) [reply tsconfig.json:16]: Done

File: package.json [hunk 11]
```diff
@@ -6,22 +6,17 @@
         "npm": ">=9",
         "node": ">=18"
     },
-    "exports": "es/index.js",
-    "module": "es/index.js",
+    "exports": "./dist/commons-ui.js",
+    "module": "./dist/commons-ui.js",
```
[the remark this answers is not in the document]
Comment (flomillot) [reply package.json:10]: We keep previous behaviour where only imports from root are allowed

[the remark this answers is not in the document]
Comment (flomillot) [reply package.json:10]: Will do an other PR

File: vite.config.mts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+ * Copyright (c) 2024, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import react from '@vitejs/plugin-react';
+import type { PluginOption } from 'vite';
+import { defineConfig } from 'vite';
+import * as path from 'path';
```
Comment (Tristan-WorkGH) [inline vite.config.mts:11]:
```suggestion
import * as path from 'node:path';
```

File: .babelrc.json [hunk 13]
```diff
@@ -0,0 +1,6 @@
+{
+  "presets": [
+    "@babel/preset-env",
+    ["@babel/preset-react", { "runtime": "automatic" }]
+  ]
+}
```
Comment (Tristan-WorkGH) [inline .babelrc.json:6]: missing new line
```suggestion
}

```

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]

File: vite.config.mts (hunk 12 above)
Comment (flomillot) [reply vite.config.mts]: Done

File: .babelrc.json (hunk 13 above)
Comment (flomillot) [reply .babelrc.json]: DOne

File: package.json [hunk 14]
```diff
@@ -58,33 +56,39 @@
...
         "@react-hook/window-size": "^3.1.1",
         "react-resizable": "^3.0.4",
+        "vite": "^5.0.11",
+        "vite-plugin-dts": "^3.7.1",
+        "vite-plugin-eslint": "^1.8.1",
+        "vite-plugin-externalize-deps": "^0.8.0",
+        "vite-plugin-svgr": "^4.2.0",
         "yup": "^1.0.0"
     },
     "eslintConfig": {
```
[the remark this answers is not in the document]
Comment (flomillot) [reply package.json:95]: I don't see it in the readme, where is it ?

**Pull request merged**